tests/unit/test_base_client.py

```diff
@@ -240,6 +258,161 @@ def _reraise_token_error(_: GlobusApp, error: TokenValidationError):
+    },
+)
+
+
+def test_app_can_redrive_gares():
+    """
+    When enabled:
+    If an app-registered client encounters a GARE-compatible 403 http response, it
+    should "redrive" it by:
+    1. Running a login flow with the gare-included authorization
+    2. Re-attempting the original request with the new token(s)
+    """
+    transfer_rs = TransferClient.resource_server
+    configurator = GlobusAppConfigurator(
+        # Start with a valid Transfer:all token
+        starting_tokens={transfer_rs: [TransferScopes.all]},
+        # On login, receive both Transfer:all and the gare-required scope
+        login_tokens={transfer_rs: [TransferScopes.all, Scope("my-cool-new-scope")]},
+    )
+    config = configurator.config(auto_redrive_gares=True)
+
+    app = UserApp(client_id="client_id", config=config)
+    transfer = TransferClient(app=app)
+
+    # Set up first a GARE response, then a successful response (on retry)
+    _GET_TASK_GARE_RESPONSE.add()
+    _GET_TASK_SUCCESS_RESPONSE.add()
+    assert transfer.get_task("foobar").http_status == 200
+    assert configurator.login_flow_manager.login_count == 1
+
+
+def test_app_gare_redriving_is_disabled_by_default():
+    transfer_rs = TransferClient.resource_server
+    configurator = GlobusAppConfigurator(
+        # Start with a valid Transfer:all token
+        starting_tokens={transfer_rs: [TransferScopes.all]},
+        # On login, receive both Transfer:all and the gare-required scope
+        login_tokens={transfer_rs: [TransferScopes.all, Scope("my-cool-new-scope")]},
+    )
+    # Don't override the `auto_redrive_gares` default of False
+    config = configurator.config()
+
+    app = UserApp(client_id="client_id", config=config)
+    transfer = TransferClient(app=app)
+
+    # Set up first a GARE response, then a successful response (on retry)
+    _GET_TASK_GARE_RESPONSE.add()
+    _GET_TASK_SUCCESS_RESPONSE.add()
+    with pytest.raises(globus_sdk.GlobusAPIError):
+        transfer.get_task("foobar")
+    assert configurator.login_flow_manager.login_count == 0
+
+
+def test_app_gare_redrive_only_occurs_once_per_request():
+    transfer_rs = TransferClient.resource_server
+    configurator = GlobusAppConfigurator(
+        # Start with a valid Transfer:all token
+        starting_tokens={transfer_rs: [TransferScopes.all]},
+        # On login, receive Transfer:all and a random scope (not required by the gare).
+        login_tokens={transfer_rs: [TransferScopes.all, Scope("my-stupid-new-scope")]},
+    )
+    config = configurator.config(auto_redrive_gares=True)
+
+    app = UserApp(client_id="client_id", config=config)
+    transfer = TransferClient(app=app)
+
+    # Set up exclusively a GARE response which will retry indefinitely.
+    _GET_TASK_GARE_RESPONSE.add()
+    with pytest.raises(globus_sdk.GlobusAPIError):
+        transfer.get_task("foobar")
+    assert configurator.login_flow_manager.login_count == 1
+
+
```

Comment:
I like all of these tests, but I feel that they're in the wrong test module. (This module is already a little bit of a hodgepodge, so that's not a new issue...)
If I were _looking for_ these tests, let's say to add another one, I wouldn't think to look in `test_base_client.py`, even though we are testing the integration between these components.

Could we maybe create a new module in `tests/functional/globus_app/` for these? Or maybe `tests/unit/globus_app/`?

`functional` vs `unit` tests are a bit blurry for the SDK, so I'm less worried about that distinction than I am about "tests of the client" vs "tests of the app".